extensions/android-paging3/src/main/java/com/squareup/sqldelight/android/paging3/OffsetQueryPagingSource.kt

```diff
@@ -34,9 +34,13 @@ internal class OffsetQueryPagingSource<RowType : Any>(
     params: LoadParams<Long>
   ): LoadResult<Long, RowType> = withContext(dispatcher) {
     try {
-      val key = params.key ?: 0L
       transacter.transactionWithResult {
         val count = countQuery.executeAsOne()
```

Comment:
Btw, it is expensive to use a transaction on every load because in sqlite (I assume the default for main use-case for sqldelight) it blocks all read/write to any table, so what we do in Room is cache this count and only get it for initial load. Any write into the DB should invalidate anyway so this count should never change while this PagingSource is able to return valid results.

Reply:
Good idea.

Reply:
On the topic of invalidation, although you use a transaction here, it is possible for a write to happen after `PagingSource.load` is called, but before you begin your transaction, and in the scenario where you return `LoadResult.Page` here before invalidation is able to propagate cancellation on this generation, it is possible you will return inconsistent results. Starting from the next Paging release (3.1.0-alpha03), we'll be adding a `LoadResult.Invalid` return type you can use to get Paging to discard your `LoadResult`. You will want to check if you are about to be invalidated after you finish loading from DB but before you return and return `LoadResult.Invalid` instead.

Reply:
This is very complicated so here is a more concrete example:

1. DB has items [1, 2, 3, 4]
2. PagingSource_1 loads [1, 2] and Paging displays those two items
3. A write to DB is kicked off to insert item 0
3. Paging library asks PagingSource_1 for an APPEND
4. Your load starts a transaction for the APPEND which needs to wait for the write to finish
5. Write finishes, DB now has items [0, 1, 2, 3, 4]
6. PagingSource_1 loads with `params.key == 2`, this results in LoadResult.Page(data = [2, 3])
7. Paging presents items [1, 2, 2, 3], duplicating one of the rows.
8. Invalidation callback finally triggers cancellation and a new generation.
9. Now PagingSource_2 is created to load data from scratch.

While you wait for PagingSource_2 to finish loading and finally send to RV to get presented, you have the duplicate row shown to UI incorrectly.

Note that it doesn't matter when you trigger invalidation, even if you are able to synchronously do it, I don't think it's possible to guarantee that cancellation successfully propagates before load completes and sends it off to presenter to show the accidentally duplicated row.

Reply:
I factored out the transaction like you mentioned. I think we'll wait until the new API in paging 3.1 is released before we worry about the complicated invalidation.